src/lsm/groove.zig

```diff
@@ -642,34 +642,32 @@ pub fn GrooveType(
             lookup_id: IdTree.LookupContext = undefined,
             lookup_object: ObjectTree.LookupContext = undefined,
 
-            /// Returns true if asynchronous I/O has been started.
-            /// Returns false if there are no more IDs to prefetch.
-            fn lookup_start(worker: *PrefetchWorker) bool {
-                const groove = worker.context.groove;
-
+            fn lookup_start_next(worker: *PrefetchWorker) void {
                 const id = worker.context.id_iterator.next() orelse {
-                    groove.prefetch_ids.clearRetainingCapacity();
-                    assert(groove.prefetch_ids.count() == 0);
-                    worker.context.worker_finished();
-                    return false;
+                    // Since worker_finished() may cause the entire prefetch to finish and call
+                    // the provided callback, we must not set worker to undefined after calling
+                    // worker_finished() as the memory may already be used for something else.
+                    const context = worker.context;
+                    worker.* = undefined;
+                    context.worker_finished();
```

Comment:
Let's go one step further still and not let any descendant node in a call graph erase it's own memory at all.

i.e. Just do `worker.context.worker_finished();` here with no `worker.* = undefined;`.

Rationale:

1. The process that kicks off all the concurrent workers is responsible for initializing their memory, however, at present, that same memory is being set to `undefined` not at the same level of concurrency control, but one node deeper in the call graph. This is rather the responsibility of whoever owns the memory, so as not to violate encapsulation.

2. This is data plane—we don't want to do memcopies for every id/object.

3. It's going to simplify callback call sites. From 3 lines down to 1. This “copy callback“ and then set to undefined pattern can be unwieldy and subtly tricky to get right. I want us to start moving away from it. For example, on line 632, we have another potential future issue, where `finish()` sets `context.* = undefined` before calling `callback(context)` passing `context` which is now `undefined`—it may be that in future `callback()` still wants to access something of `context`, rather than only using it for `@fieldParentPtr()`, and this is especially likely if the memory is managed at a higher layer.

Reply:
Yeah, I agree that this pattern doesn't seem to be worth the tradeoff in practice. Added a commit removing all instances of it across our code.

(Aside: what do you think of the "style" commit message prefix for this kind of thing?)

Reply:
Great, thanks for removing all the other instances.

I would rather not introduce commit message prefixes beyond the functional namespaces (e.g. `LSM`, `VSR` etc). Code is style/design (safety/performance/experience) and design is code. It's hard to separate those things. They're so closely joined into the conceptual integrity of what makes TB, TB.

Reply:
Makes sense to me, switched to `LSM,VSR` for the prefix :+1: